### PR TITLE
Add rule : Property unit whitelist

### PR DESCRIPTION
**Pull request by makotot**
Ref : https://github.com/stylelint/stylelint/issues/430

### CODE REVIEW

File: src/rules/property-unit-whitelist/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
+# property-unit-whitelist
+
+Specify a whitelist of disallowed units for specific properties.
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/README.md]: Should be "allowed" I believe.

File: src/rules/property-unit-whitelist/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+Given:
+
+```js
+{
+  "font-size": ["em", "px"],
+  "animation": ["s"]
+}
+```
+
+The following patterns are considered warnings:
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/README.md]: Shouldn't these examples be reversed?

File: src/rules/property-unit-whitelist/index.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    root.walkDecls(decl => {
+
+      const { prop, value } = decl
+      const whitelistInProp = whitelist[prop]
+
+      valueParser(value).walk(function (node) {
+        const unit = valueParser.unit(node.value).unit
+
+        if (whitelistInProp && unit && whitelistInProp.indexOf(unit) === -1 && node.type !== "string") {
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/index.js]: This is long enough that I'd split it up to be more readable. Maybe something like:

```js
if (
  whitelistInProp 
  && unit 
  && whitelistInProp.indexOf(unit) === -1 
  && node.type !== "string"
) {
```

File: src/rules/property-unit-whitelist/index.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      actual: whitelist,
+      possible: [isObject],
+    })
+    if (!validOptions) { return }
+
+    root.walkDecls(decl => {
+
+      const { prop, value } = decl
+      const whitelistInProp = whitelist[prop]
+
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/index.js]: This variable should be named `propWhitelist`.

And at this point you can early return if it is undefined, right?

```js
const propWhitelist = whitelist[prop]
if (!propWhitelist) { return }
```

File: src/rules/property-unit-whitelist/index.js (hunk 3 above)
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/index.js]: The `whitelistInProp` can be done with an early return above, as described in the other comment.

File: src/rules/property-unit-whitelist/__tests__/index.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  tr.notOk("a { margin: calc(10em - 10px); }", {
+    message: messages.rejected("margin", "px"),
+    column: 25,
+  })
+
+  tr.notOk("a { animation: animation-name 300ms ease; }", {
+    message: messages.rejected("animation", "ms"),
+    column: 31,
+  })
+})
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/__tests__/index.js]: How does this handle vendor-prefixes?

**Comment by makotot** [comment]
@davidtheclark  Thanks :) Updated this PR.

File: src/rules/property-unit-whitelist/index.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      const { prop, value } = decl
+      const propWhitelist = whitelist[prop]
+
+      if (!propWhitelist) { return }
+
+      valueParser(value).walk(function (node) {
+        const unit = valueParser.unit(node.value).unit
+
+        if (unit && propWhitelist.indexOf(unit) === -1 && node.type !== "string") {
```
Comment (davidtheclark) [inline src/rules/property-unit-whitelist/index.js]: Sorry but I thought of one other easy optimization here. It probably applies to the other whitelist/blacklist rules, as well, if you feel like retrofitting ...

So the `node.type` check can go before the use of `valueParser`, thereby saving the parsing step if the node is a string. 

```js
valueParser(value).walk(function (node) {
  if (node.type === "string") { return }
  const unit = valueParser.unit(node.value).unit
```

**Comment by jeddy3** [comment]
Excellent to see the last of the `blacklist` and `whitelist` rules done! Thanks @makotot 

It's looking great. I have one query though: one of the original use cases I had in mind for this rule was specifying _no_ units for the `line-height` property e.g. `{ "line-height": [] }`. After a quick glance the code looks like it will support this, but can you please add a couple of tests in to make sure?

**Comment by davidtheclark** [comment]
@jeddy3 : Yes, I agree that this rule should be able to enforce no units on line-height.

**Comment by davidtheclark** [comment]
@jeddy3 : After this feature is finish do you agree we're ready for a 2.1.0 release?

**Comment by jeddy3** [comment]
@davidtheclark Defo :)

**Comment by makotot** [comment]
@davidtheclark @jeddy3  Thanks! Updated PR :)
> So the node.type check can go before the use of valueParser, thereby saving the parsing step if the node is a string.

Is it better to modify [property-unit-blacklist](https://github.com/stylelint/stylelint/blob/master/src/rules/property-unit-blacklist/index.js#L33) in same way?

**Comment by davidtheclark** [comment]
@makotot : Excellent. I'll merge this. If you update property-unit-blacklist that would be fantastic.

**Comment by makotot** [comment]
@davidtheclark  Thanks!
> If you update property-unit-blacklist that would be fantastic.

Okay. I'll update property-unit-blacklist :)

**Comment by jeddy3** [comment]
@makotot Thanks for adding in the `line-height` test :)

**Pull request merged**